signac/contrib/migration/__init__.py

```diff
@@ -5,102 +5,141 @@
             raise RuntimeError(
-                "The signac schema version used by this project is {}, but signac {} "
-                "uses schema version {} and does not know how to migrate.".format(
-                    get_config_schema_version(), __version__, schema_version
-                )
+                "The signac schema version used by this project is "
+                f"{_get_config_schema_version(root_directory, guess)}, but "
+                f"signac {__version__} uses schema version {schema_version} "
+                "and does not know how to migrate."
             )
 
 
-def apply_migrations(project):
-    """Apply migrations to a project."""
-    with _lock_for_migration(project):
-        for (origin, destination), migrate in _collect_migrations(project):
-            try:
-                print(
-                    f"Applying migration for version {origin} to {destination}... ",
-                    end="",
-                    file=sys.stderr,
-                )
-                migrate(project)
-            except Exception as e:
-                raise RuntimeError(f"Failed to apply migration {destination}.") from e
-            else:
-                _update_project_config(project, schema_version=destination)
-                print("OK", file=sys.stderr)
-                yield origin, destination
+def apply_migrations(root_directory):
+    """Apply migrations to a project.
+
+    This function identifies and performs all the necessary schema migrations
+    to bring a project up to date with the current schema version of signac.
+    The calling code does not require prior knowledge of the schema version of
+    the project, and the function is idempotent when applied to projects that
+    already have an up-to-date schema.
+
+    Parameters
+    ----------
+    root_directory : str
+        The path to the project to migrate.
+    """
+    try:
+        lock = FileLock(os.path.join(root_directory, FN_MIGRATION_LOCKFILE))
+        with lock:
+            for (origin, destination), migrate in _collect_migrations(root_directory):
+                try:
+                    print(
+                        f"Applying migration for version {origin} to {destination}... ",
+                        end="",
+                        file=sys.stderr,
+                    )
+                    migrate(root_directory)
+                except Exception as e:
+                    raise RuntimeError(
+                        f"Failed to apply migration {destination}."
+                    ) from e
+                else:
+                    config = _CONFIG_LOADERS[version.parse(destination).public](
+                        root_directory
+                    )
+                    config["schema_version"] = destination
+                    config.write()
+
+                    print("OK", file=sys.stderr)
+    finally:
+        try:
+            os.unlink(lock.lock_file)
+        except FileNotFoundError:
+            pass
```

Comment:
Is this not the responsibility of the `FileLock` context manager protocol?

Reply:
I would think so, but I assumed there was a reason this code was added in the first place. If not, I'm happy to remove this whole try-except block and simplify the code.

Reply:
🤷

Reply:
@bdice do you know?

Reply:
I have since realized that the context manager does in fact not remove the lock file. The lock file is "unlocked", but not necessarily removed.

Reply:
Correct. I tried it locally. The `FileLock` context manager does not remove the lock file -- it creates it and uses it. The `SoftFileLock`, which only checks file existence and does not rely on the deeper OS primitives for file locking, performs cleanup and removes the file when exiting the context manager. https://py-filelock.readthedocs.io/en/latest/index.html#filelock-vs-softfilelock